In [1]:
# Checking python version
from platform import python_version
print(python_version())

3.9.7


In [2]:
# Import packages
import pandas as pd # for data analysis
import numpy as np # for mathematical and logical operations

import seaborn as sns # for Data visualization
import matplotlib.pyplot as plt # for Data visualization
import matplotlib
%matplotlib inline
matplotlib.rcParams['figure.figsize'] = (10,6)

In [3]:
# Checking % of missing values in columns
def check_missing_values(df):
    for col in df.columns:
        pct_missing = np.mean(df[col].isnull())
        print('{} - {}%'.format(col, round(pct_missing*100)))

In [4]:
# Loading the dataset
df = pd.read_csv("Prescribed_Medication_DataScientistTask_2023_ForCandidates.csv")
df.head(5)

ID Source        Date  \
0   23.0   Q11B  10/11/2004   
1   86.0   Q11B  04/05/2005   
2   86.0   Q17D  19/12/2007   
3   86.0    Q29     9.12.14   
4  157.0    Q29    13.11.14   

                                      Input_original  \
0                                        SIMVASTATIN   
1  BENDROFLUAZIDE;GLYCERYL TRINITRATE SPRAY;LISIN...   
2  LISINOPRIL;HYPROMELLOSE EYE DROPS;BENDROFLUMET...   
3  ATENOLOL~LANSOPRAZOLE~LISINOPRIL~DISPERSIBLE A...   
4  RAMIPRIL 10MG~PREDNISOLONE 2.5MG~NITROFURANTOI...   

                                             Product  \
0                                        Simvastatin   
1  Bendroflumethiazide;Glyceryl Trinitrate;Lisino...   
2  Lisinopril;Hypromellose;Bendroflumethiazide;At...   
3  Aspirin;Atenolol;Bisacodyl;Glyceryl Trinitrate...   
4               Nitrofurantoin;Prednisolone;Ramipril   

                                               Class  
0                             Lipid-Regulating Drugs  
1  Thiazides And Related Diuretics;Management of ...  
2  Angiotensin-Converting Enzyme Inhibitors;Tear ...  
3  Antiplatelet Drugs;Beta-Adrenoceptor Blocking ...  
4  Urinary-Tract Infections;Corticosteroids;Angio...

# EDA

In [5]:
# Checking number of record(rows) and variable(cols)
df.shape

(1345, 6)

In [6]:
# Show the datatypes
df.dtypes

ID                float64
Source             object
Date               object
Input_original     object
Product            object
Class              object
dtype: object

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1345 entries, 0 to 1344
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   ID              1345 non-null   float64
 1   Source          1345 non-null   object 
 2   Date            1345 non-null   object 
 3   Input_original  1344 non-null   object 
 4   Product         1345 non-null   object 
 5   Class           1345 non-null   object 
dtypes: float64(1), object(5)
memory usage: 63.2+ KB


In [8]:
check_missing_values(df)

ID - 0%
Source - 0%
Date - 0%
Input_original - 0%
Product - 0%
Class - 0%


In [9]:
import warnings
warnings.filterwarnings('ignore')
# Converting date to dtype datetime64
df['Date'] = pd.to_datetime(df.Date)
df['Date'].dt.strftime('%d-%m-%Y')

0       11-10-2004
1       05-04-2005
2       19-12-2007
3       12-09-2014
4       13-11-2014
           ...    
1340    28-02-2019
1341    09-04-2019
1342    24-04-2019
1343    16-04-2019
1344    17-04-2019
Name: Date, Length: 1345, dtype: object

In [10]:
# Converting float64 to Int64 for ID column
df['ID'] = df['ID'].astype('Int64') 

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1345 entries, 0 to 1344
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   ID              1345 non-null   Int64         
 1   Source          1345 non-null   object        
 2   Date            1345 non-null   datetime64[ns]
 3   Input_original  1344 non-null   object        
 4   Product         1345 non-null   object        
 5   Class           1345 non-null   object        
dtypes: Int64(1), datetime64[ns](1), object(4)
memory usage: 64.5+ KB


# Task A

In [12]:
# Splitting the Class variable as different products
df[['C0', 'C1', 'C2', 'C3', 'C4', 'C5', 'C6', 'C7', 'C8', 'C9', 'C10', 'C11', 'C12', 'C13', 'C14', 'C15', 'C16']] = df['Class'].str.split(';', expand=True)
df.head(5)

ID Source       Date                                     Input_original  \
0   23   Q11B 2004-10-11                                        SIMVASTATIN   
1   86   Q11B 2005-04-05  BENDROFLUAZIDE;GLYCERYL TRINITRATE SPRAY;LISIN...   
2   86   Q17D 2007-12-19  LISINOPRIL;HYPROMELLOSE EYE DROPS;BENDROFLUMET...   
3   86    Q29 2014-09-12  ATENOLOL~LANSOPRAZOLE~LISINOPRIL~DISPERSIBLE A...   
4  157    Q29 2014-11-13  RAMIPRIL 10MG~PREDNISOLONE 2.5MG~NITROFURANTOI...   

                                             Product  \
0                                        Simvastatin   
1  Bendroflumethiazide;Glyceryl Trinitrate;Lisino...   
2  Lisinopril;Hypromellose;Bendroflumethiazide;At...   
3  Aspirin;Atenolol;Bisacodyl;Glyceryl Trinitrate...   
4               Nitrofurantoin;Prednisolone;Ramipril   

                                               Class  \
0                             Lipid-Regulating Drugs   
1  Thiazides And Related Diuretics;Management of ...   
2  Angiotensin-Converting Enzyme Inhibitors;Tear ...   
3  Antiplatelet Drugs;Beta-Adrenoceptor Blocking ...   
4  Urinary-Tract Infections;Corticosteroids;Angio...   

                                         C0  \
0                    Lipid-Regulating Drugs   
1           Thiazides And Related Diuretics   
2  Angiotensin-Converting Enzyme Inhibitors   
3                        Antiplatelet Drugs   
4                  Urinary-Tract Infections   

                                         C1  \
0                                      None   
1               Management of Anal Fissures   
2  Tear Deficiency,Eye Lubricant/Astringent   
3          Beta-Adrenoceptor Blocking Drugs   
4                           Corticosteroids   

                                         C2                                C3  \
0                                      None                              None   
1  Angiotensin-Converting Enzyme Inhibitors            Proton Pump Inhibitors   
2           Thiazides And Related Diuretics  Beta-Adrenoceptor Blocking Drugs   
3                       Stimulant Laxatives       Management of Anal Fissures   
4  Angiotensin-Converting Enzyme Inhibitors                              None   

   ...                                        C7  \
0  ...                                      None   
1  ...                                      None   
2  ...   Non-Opioid Analgesics And Compound Prep   
3  ...  Angiotensin-Converting Enzyme Inhibitors   
4  ...                                      None   

                                        C8  \
0                                     None   
1                                     None   
2       Other Drugs For Rheumatic Diseases   
3  Non-Opioid Analgesics And Compound Prep   
4                                     None   

                                         C9         C10   C11   C12   C13  \
0                                      None        None  None  None  None   
1                                      None        None  None  None  None   
2               Management of Anal Fissures        None  None  None  None   
3  Tear Deficiency,Eye Lubricant/Astringent  Emollients  None  None  None   
4                                      None        None  None  None  None   

    C14   C15   C16  
0  None  None  None  
1  None  None  None  
2  None  None  None  
3  None  None  None  
4  None  None  None  

[5 rows x 23 columns]

In [13]:
check_missing_values(df)

ID - 0%
Source - 0%
Date - 0%
Input_original - 0%
Product - 0%
Class - 0%
C0 - 0%
C1 - 56%
C2 - 70%
C3 - 80%
C4 - 86%
C5 - 90%
C6 - 92%
C7 - 94%
C8 - 96%
C9 - 97%
C10 - 98%
C11 - 99%
C12 - 100%
C13 - 100%
C14 - 100%
C15 - 100%
C16 - 100%


In [14]:
# Dropping the Class column and the columns with 100% missing values
df = df.drop('Class', axis=1)
df.drop(columns =["C12", "C13", "C14", "C15", "C15"], inplace = True)

In [15]:
# Selecting the required columns only
taskA_df = df[['ID', 'Date', 'Source', 'C0', 'C1', 'C2', 'C3', 'C4', 'C5', 'C6', 'C7', 'C8', 'C9', 'C10', 'C11']]
taskA_df.head(5)

ID       Date Source                                        C0  \
0   23 2004-10-11   Q11B                    Lipid-Regulating Drugs   
1   86 2005-04-05   Q11B           Thiazides And Related Diuretics   
2   86 2007-12-19   Q17D  Angiotensin-Converting Enzyme Inhibitors   
3   86 2014-09-12    Q29                        Antiplatelet Drugs   
4  157 2014-11-13    Q29                  Urinary-Tract Infections   

                                         C1  \
0                                      None   
1               Management of Anal Fissures   
2  Tear Deficiency,Eye Lubricant/Astringent   
3          Beta-Adrenoceptor Blocking Drugs   
4                           Corticosteroids   

                                         C2                                C3  \
0                                      None                              None   
1  Angiotensin-Converting Enzyme Inhibitors            Proton Pump Inhibitors   
2           Thiazides And Related Diuretics  Beta-Adrenoceptor Blocking Drugs   
3                       Stimulant Laxatives       Management of Anal Fissures   
4  Angiotensin-Converting Enzyme Inhibitors                              None   

                                         C4                                C5  \
0                                      None                              None   
1                    Lipid-Regulating Drugs  Beta-Adrenoceptor Blocking Drugs   
2                        Antiplatelet Drugs            Proton Pump Inhibitors   
3  Tear Deficiency,Eye Lubricant/Astringent                 Osmotic Laxatives   
4                                      None                              None   

                                 C6                                        C7  \
0                              None                                      None   
1  Beta-Adrenoceptor Blocking Drugs                                      None   
2            Lipid-Regulating Drugs   Non-Opioid Analgesics And Compound Prep   
3            Proton Pump Inhibitors  Angiotensin-Converting Enzyme Inhibitors   
4                              None                                      None   

                                        C8  \
0                                     None   
1                                     None   
2       Other Drugs For Rheumatic Diseases   
3  Non-Opioid Analgesics And Compound Prep   
4                                     None   

                                         C9         C10   C11  
0                                      None        None  None  
1                                      None        None  None  
2               Management of Anal Fissures        None  None  
3  Tear Deficiency,Eye Lubricant/Astringent  Emollients  None  
4                                      None        None  None

In [16]:
# Saving filtered data to csv
taskA_df.to_csv("taskA_dataset.csv", index= False)

Task A - notes

In the given dataset, there are 1345 records with 6 columns. There are no missing values(emply records are filled with 999xxxx) in each column. For proper data processing, ID and Date columns are converted as Int and Datetime format respectively.
For facilitating the further data processing, I did the
- splitting the Class variable as separated columns and
- converting Date column with the format of %d-%m-%Y, then
- saved the new dataset as a taskA_dataste.csv file.
For further data analysis, it will be easy to use the Date variable and different types of separated classes. 

# Task B

In [17]:
# row and col counts in taskA_dataset 
taskA_df.shape

(1345, 15)

### Notes
Filter the data rows from the medication dataset till the year of June 2017 to merge with measurements of cholesterol levels   which were taken in June 2017
- filter out the dataset with the "Date" till June 2017
- filter out the participants with “Lipid-Regulating Drugs” class

In [18]:
# filter out the dataset with the "Date" till 31 June 2017
df_by_07_17 = df[(df['Date'] <= "2017-07-31")]
df_by_07_17.shape

(1252, 18)

In [19]:
# Filter out the participants with product name/class “Lipid-Regulating Drugs”
columns = ['ID', 'Date', 'Source', 'C0', 'C1', 'C2', 'C3', 'C4', 'C5', 'C6', 'C7', 'C8', 'C9', 'C10', 'C11']
taskB_dataset = df_by_07_17[df_by_07_17['C0'].isin(["Lipid-Regulating Drugs"]) 
   | df['C1'].isin(["Lipid-Regulating Drugs"])
  | df['C2'].isin(["Lipid-Regulating Drugs"])
  | df['C3'].isin(["Lipid-Regulating Drugs"])
  | df['C4'].isin(["Lipid-Regulating Drugs"])
  | df['C5'].isin(["Lipid-Regulating Drugs"])
  | df['C6'].isin(["Lipid-Regulating Drugs"])
  | df['C7'].isin(["Lipid-Regulating Drugs"])
  | df['C8'].isin(["Lipid-Regulating Drugs"])
  | df['C9'].isin(["Lipid-Regulating Drugs"])
  | df['C10'].isin(["Lipid-Regulating Drugs"])
  | df['C11'].isin(["Lipid-Regulating Drugs"])][columns]

In [20]:
# row and col counts in taskB_dataset 
taskB_dataset.shape

(187, 15)

#### Notes
There are more than one row of participant data could include in this subset of taskB dataset. 

Please check the following code script. There are two rows for participant ID-86.

Therefore, the total number of participants with "Lipid-Regulating Drugs" class as of as of June 2017 could be less than the data row of 187.

In [21]:
# Checking the multiple rows for a particular ID
taskB_dataset[(taskB_dataset["ID"]==86)]

ID       Date Source                                        C0  \
1  86 2005-04-05   Q11B           Thiazides And Related Diuretics   
2  86 2007-12-19   Q17D  Angiotensin-Converting Enzyme Inhibitors   

                                         C1  \
1               Management of Anal Fissures   
2  Tear Deficiency,Eye Lubricant/Astringent   

                                         C2                                C3  \
1  Angiotensin-Converting Enzyme Inhibitors            Proton Pump Inhibitors   
2           Thiazides And Related Diuretics  Beta-Adrenoceptor Blocking Drugs   

                       C4                                C5  \
1  Lipid-Regulating Drugs  Beta-Adrenoceptor Blocking Drugs   
2      Antiplatelet Drugs            Proton Pump Inhibitors   

                                 C6                                       C7  \
1  Beta-Adrenoceptor Blocking Drugs                                     None   
2            Lipid-Regulating Drugs  Non-Opioid Analgesics And Compound Prep   

                                   C8                           C9   C10   C11  
1                                None                         None  None  None  
2  Other Drugs For Rheumatic Diseases  Management of Anal Fissures  None  None

In [22]:
# Save filtered data to csv
taskB_dataset.to_csv("taskB_dataset.csv", index= False)

# Task C

- What some of the limitations of this data might be? 

Input_original and Product columns say the same/similar information.  When we build further statistical or ML algorithms with the dataset, it assumes that the variables(columns) should be independent of each other to get better accuracy and less model complexity.

After we have processed Input_original column, it should be removed not for having redundant data in the dataset for further data analysis.

- Any ways that you could deal with missing or incomplete data? 

For those rows which did not include any useful information, those should be removed before further processing. Please check the following code script, there are 424 rows of data in total with no useful information as of June 2017.

- Anything else it is important to consider when using this data?  

After we have processed Input_original and Product columns, it assumes that the dataset is good to go for further data analysis. 

In [23]:
# Checking how many records in which the participants did not give any medications as of June 2017
df_by_07_17[(df_by_07_17["Input_original"] == "999906")]

ID Source       Date Input_original Product      C0    C1    C2  \
18      482    Q29 2016-08-25         999906  999906  999906  None  None   
19      497   Q11B 2005-06-23         999906  999906  999906  None  None   
20      497   Q17D 2009-09-03         999906  999906  999906  None  None   
21      497   Q11B 2014-04-14         999906  999906  999906  None  None   
22      497    Q29 2014-12-17         999906  999906  999906  None  None   
...     ...    ...        ...            ...     ...     ...   ...   ...   
1309  20766   Q17D 2009-02-18         999906  999906  999906  None  None   
1312  20812   Q11B 2005-07-15         999906  999906  999906  None  None   
1318  20884   Q11B 2004-11-23         999906  999906  999906  None  None   
1319  20884    Q29 2014-10-16         999906  999906  999906  None  None   
1320  20884    Q36 2017-06-21         999906  999906  999906  None  None   

        C3    C4    C5    C6    C7    C8    C9   C10   C11   C16  
18    None  None  None  None  None  None  None  None  None  None  
19    None  None  None  None  None  None  None  None  None  None  
20    None  None  None  None  None  None  None  None  None  None  
21    None  None  None  None  None  None  None  None  None  None  
22    None  None  None  None  None  None  None  None  None  None  
...    ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  
1309  None  None  None  None  None  None  None  None  None  None  
1312  None  None  None  None  None  None  None  None  None  None  
1318  None  None  None  None  None  None  None  None  None  None  
1319  None  None  None  None  None  None  None  None  None  None  
1320  None  None  None  None  None  None  None  None  None  None  

[424 rows x 18 columns]